goagen/gen_test/generator.go

```diff
@@ -0,0 +1,299 @@
+					} else {
+						for _, view := range mediaType.Views {
+							methods = append(methods, createTestMethod(res, action, response, route, routeIndex, mediaType, view))
+						}
+					}
+				}
+			}
+		}
+
+		g.genfiles = append(g.genfiles, filename)
+		err = resourceTestContextTmpl.Execute(file, methods)
+		if err != nil {
+			panic(err)
+		}
+		return file.FormatCode()
+	})
+}
+
+func createTestMethod(resource *design.ResourceDefinition, action *design.ActionDefinition, response *design.ResponseDefinition, route *design.RouteDefinition, routeIndex int, mediaType *design.MediaTypeDefinition, view *design.ViewDefinition) TestMethod {
+	routeNameQualifier := suffixRoute(action.Routes, routeIndex)
+	viewNameQualifier := func() string {
+		if view != nil && view.Name != "default" {
+			return view.Name
+		}
+		return ""
+	}()
+	method := TestMethod{}
+	method.Name = fmt.Sprintf("%s%s%s%s%s", codegen.Goify(action.Name, true), codegen.Goify(resource.Name, true), codegen.Goify(response.Name, true), routeNameQualifier, codegen.Goify(viewNameQualifier, true))
+	method.ActionName = codegen.Goify(action.Name, true)
+	method.ResourceName = codegen.Goify(resource.Name, true)
+	method.Comment = fmt.Sprintf("test setup")
+	method.ControllerName = fmt.Sprintf("%s.%sController", TargetPackage, codegen.Goify(resource.Name, true))
+	method.ContextVarName = fmt.Sprintf("%sCtx", codegen.Goify(action.Name, false))
+	method.ContextType = fmt.Sprintf("%s.New%s%sContext", TargetPackage, codegen.Goify(action.Name, true), codegen.Goify(resource.Name, true))
+	method.RouteVerb = route.Verb
+	method.Status = response.Status
+	method.FullPath = goPathFormat(route.FullPath())
+
+	if view != nil && mediaType != nil {
+		p, _, err := mediaType.Project(view.Name)
+		if err != nil {
+			panic(err)
+		}
+		tmp := fmt.Sprintf("%s.%s", TargetPackage, codegen.GoTypeName(p, nil, 0, false))
+		validate := codegen.RecursiveChecker(p.AttributeDefinition, false, false, false, "payload", "raw", 1, true)
+
+		returnType := ObjectType{}
+		returnType.Type = tmp
+		returnType.Pointer = "*"
+		returnType.Validatable = validate != ""
+
+		method.ReturnType = &returnType
+	}
+
+	if len(route.Params()) > 0 {
+		var params = []ObjectType{}
+		for _, paramName := range route.Params() {
+			for name, att := range action.Params.Type.ToObject() {
+				if name == paramName {
+					param := ObjectType{}
+					param.Name = codegen.Goify(name, false)
+					param.Type = codegen.GoTypeRef(att.Type, nil, 0, false)
+					if att.Type.IsPrimitive() && action.Params.IsPrimitivePointer(name) {
+						param.Pointer = "*"
+					}
+					params = append(params, param)
+				}
+			}
+		}
+		method.Params = params
+	}
+
+	if action.Payload != nil {
+		payload := ObjectType{}
+		payload.Name = "payload"
+		payload.Type = fmt.Sprintf("%s.%s", TargetPackage, codegen.Goify(action.Payload.TypeName, true))
+		payload.Pointer = "*"
```

Comment:
A payload may not be a struct. To be fair this is not something that has been tested explicitly so it's possible other generators don't support it but it would be nice for this new one to not have the bug from the get go :)

The design supports doing:
```go
Payload(Integer)
```
for example.